File: sample-new-resource-detector/src/test/java/nl/haploid/octowight/sample/detector/CaptainResourceDetectorTest.java
package nl.haploid.octowight.sample.detector;

import mockit.Injectable;
import mockit.Mocked;
import mockit.StrictExpectations;
import mockit.Tested;
import nl.haploid.octowight.AtomChangeEvent;
import nl.haploid.octowight.registry.data.ResourceRoot;
import nl.haploid.octowight.registry.data.ResourceRootFactory;
import nl.haploid.octowight.sample.TestData;
import nl.haploid.octowight.sample.repository.PersonDmo;
import nl.haploid.octowight.sample.repository.RoleDmo;
import nl.haploid.octowight.sample.repository.RoleDmoRepository;
import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

public class CaptainResourceDetectorTest {

    @Tested
    private CaptainResourceDetector detector;

    @Injectable
    private RoleDmoRepository roleDmoRepository;

    @Injectable
    private ResourceRootFactory resourceRootFactory;

    @Test
    public void testGetAtomTypes() throws Exception {
        assertEquals(1, detector.getAtomTypes().size());
    }

    @Test
    public void testDetect() throws Exception {
        final AtomChangeEvent event1 = TestData.atomChangeEvent();
        final AtomChangeEvent event2 = TestData.atomChangeEvent();
        final AtomChangeEvent event3 = TestData.atomChangeEvent();
        final List<AtomChangeEvent> events = Arrays.asList(event1, event2, event3);
        final Long id1 = event1.getAtomId();
        final Long id2 = event2.getAtomId();
        final RoleDmo roleDmo1 = TestData.roleDmo(id1);
        final RoleDmo roleDmo2 = TestData.roleDmo(id2);
        final HashMap<Long, RoleDmo> rolesById = new HashMap<Long, RoleDmo>() {{
            put(id1, roleDmo1);
            put(id2, roleDmo2);
        }};
        final ResourceRoot resourceRoot = TestData.resourceRoot(id2);
        final List<ResourceRoot> expectedResourceRoots = Collections.singletonList(resourceRoot);
        new StrictExpectations(detector) {{
            detector.getRolesById(events);
            times = 1;
            result = rolesById;
            detector.isCaptain(roleDmo1);
            times = 1;
            result = false;
            detector.isCaptain(roleDmo2);
            times = 1;
            result = true;
            resourceRootFactory.fromAtomChangeEvent(event2, CaptainResourceDetector.RESOURCE_TYPE);
            times = 1;
            result = resourceRoot;
        }};
        final List<ResourceRoot> actualResourceRoots = detector.detect(events);
        assertEquals(expectedResourceRoots, actualResourceRoots);
    }

    @Test
    public void testGetRolesById() throws Exception {
        final AtomChangeEvent event1 = TestData.atomChangeEvent();
        final AtomChangeEvent event2 = TestData.atomChangeEvent();
        final List<AtomChangeEvent> events = Arrays.asList(event1, event2);
        final Long id1 = event1.getAtomId();
        final Long id2 = event2.getAtomId();
        final RoleDmo roleDmo1 = TestData.roleDmo(id1);
        final RoleDmo roleDmo2 = TestData.roleDmo(id2);
        new StrictExpectations() {{
            roleDmoRepository.findAll(Arrays.asList(id1, id2));
            times = 1;
            result = Arrays.asList(roleDmo1, roleDmo2);
        }};
        final Map<Long, RoleDmo> rolesById = detector.getRolesById(events);
        assertEquals(roleDmo1, rolesById.get(id1));
        assertEquals(roleDmo2, rolesById.get(id2));
    }

    @Test
    public void testIsCaptain(final @Mocked PersonDmo personDmo) {
        final RoleDmo roleDmo1 = TestData.roleDmo(personDmo, "harpooner");
        final RoleDmo roleDmo2 = TestData.roleDmo(personDmo, CaptainResourceDetector.ROLE_TYPE);
        assertFalse(detector.isCaptain(roleDmo1));
        assertTrue(detector.isCaptain(roleDmo2));
    }
}

File: sample-new-resource-detector/src/test/scala/nl/haploid/octowight/sample/detector/CaptainResourceDetectorTest.scala
package nl.haploid.octowight.sample.detector

import java.util

import nl.haploid.octowight.registry.data.ResourceRoot
import nl.haploid.octowight.sample.data.CaptainResource
import nl.haploid.octowight.sample.repository.{PersonDmo, RoleDmoRepository}
import nl.haploid.octowight.sample.{AbstractTest, TestData}
import nl.haploid.octowight.{Mocked, Tested}
import org.easymock.EasyMock

class CaptainResourceDetectorTest extends AbstractTest {
  @Tested private[this] val detector: CaptainResourceDetector = null
  @Mocked private[this] val roleDmoRepository: RoleDmoRepository = null

  behavior of "Captain resource detector"

  it should "have an atom category" in {
    detector.atomCategories should have size 1
  }

  it should "detect captains" in {
    val detector = withMocks(EasyMock.createMockBuilder(classOf[CaptainResourceDetector])
      .addMockedMethod("findRolesById")
      .addMockedMethod("isCaptain")
      .createMock())
    val event1 = TestData.atomChangeEvent
    val event2 = TestData.atomChangeEvent
    val event3 = TestData.atomChangeEvent
    val events = List(event1, event2, event3)
    val roleDmo1 = TestData.roleDmo(event1.atomId)
    val roleDmo2 = TestData.roleDmo(event2.atomId)
    val rolesById = Map((event1.atomId, roleDmo1), (event2.atomId, roleDmo2))
    val resourceRoot = ResourceRoot(event2, CaptainResource.ResourceCollection)
    val expectedResourceRoots = List(resourceRoot)
    expecting {
      detector.findRolesById(events) andReturn rolesById once()
      detector.isCaptain(roleDmo1) andReturn false once()
      detector.isCaptain(roleDmo2) andReturn true once()
    }
    whenExecuting(detector) {
      val actualResourceRoots = detector.detect(events)
      actualResourceRoots should be(expectedResourceRoots)
    }
  }

  it should "get roles by id" in {
    val event1 = TestData.atomChangeEvent
    val event2 = TestData.atomChangeEvent
    val events = List(event1, event2)
    val id1 = event1.atomId
    val id2 = event2.atomId
    val roleDmo1 = TestData.roleDmo(id1)
    val roleDmo2 = TestData.roleDmo(id2)
    expecting {
      roleDmoRepository.findAll(util.Arrays.asList(id1, id2)) andReturn util.Arrays.asList(roleDmo1, roleDmo2) once()
    }
    whenExecuting(roleDmoRepository) {
      val rolesById = detector.findRolesById(events)
      rolesById.get(id1).orNull should be(roleDmo1)
      rolesById.get(id2).orNull should be(roleDmo2)
    }
  }

  it should "recognize a captain" in {
    val personDmo = mock[PersonDmo]
    val roleDmo1 = TestData.roleDmo(personDmo, "harpooner")
    val roleDmo2 = TestData.roleDmo(personDmo, CaptainResourceDetector.RoleType)
    detector.isCaptain(roleDmo1) should be(right = false)
    detector.isCaptain(roleDmo2) should be(right = true)
  }
}
